Sources/Environment/local_environment.cpp

```diff
@@ -0,0 +1,354 @@
+    double dist;
+
+    // Record the distance and position of every image in the cutoff sphere.
+    for (int n = 0; n < noa; n++){
+        diff_curr = structure.wrapped_positions.row(n) - pos_atom;
+        for (int s1 = -sweep_val; s1 < sweep_val + 1; s1++){
+            for (int s2 = -sweep_val; s2 < sweep_val + 1; s2++){
+                for (int s3 = -sweep_val; s3 < sweep_val + 1; s3++){
+                    im = diff_curr + s1 * vec1 + s2 * vec2 + s3 * vec3;
+                    dist = sqrt(im(0) * im(0) + im(1) * im(1) + im(2) * im(2));
+                    if ((dist < cutoff) && (dist != 0)){
+                        dists[counter] = dist;
+                        xvals[counter] = im(0);
+                        yvals[counter] = im(1);
+                        zvals[counter] = im(2);
+                        cutoff_count ++;
+                    }
+                    counter ++;
+                }
+            }
+        }
+    }
+
+    // Store information about atoms in the cutoff sphere.
+    environment_indices.resize(cutoff_count);
+    environment_species.resize(cutoff_count);
+    rs.resize(cutoff_count);
+    xs.resize(cutoff_count);
+    ys.resize(cutoff_count);
+    zs.resize(cutoff_count);
+    xrel.resize(cutoff_count);
+    yrel.resize(cutoff_count);
+    zrel.resize(cutoff_count);
+    int spec_curr, unique_check;
+    double dist_curr, xcurr, ycurr, zcurr;
+    int bond_count = 0;
+    counter = 0;
+
+    for (int m = 0; m < noa; m++){
+        spec_curr = structure.coded_species[m];
+
+        // Add central atom to the neighbor list.
+        unique_check = 0;
+        if (m == atom){
+            neighbor_list.push_back(m);
+            unique_check = 1;
+        }
+
+        for (int n = 0; n < sweep_no; n++){
+            dist_curr = dists[counter];
+            if ((dist_curr < cutoff) && (dist_curr != 0)){
+                environment_indices[bond_count] = m;
+                environment_species[bond_count] = spec_curr;
+
+                xcurr = xvals[counter];
+                ycurr = yvals[counter];
+                zcurr = zvals[counter];
+
+                rs[bond_count] = dist_curr;
+                xs[bond_count] = xcurr;
+                ys[bond_count] = ycurr;
+                zs[bond_count] = zcurr;
+                xrel[bond_count] = xcurr / dist_curr;
+                yrel[bond_count] = ycurr / dist_curr;
+                zrel[bond_count] = zcurr / dist_curr;
+
+                bond_count ++;
+
+                // Update neighbor list.
+                if (unique_check == 0){
+                    neighbor_list.push_back(m);
+                    unique_check = 1;
+                }
+            }
+            counter ++;
+        }
+    }
+
+    delete [] dists; delete [] xvals; delete [] yvals; delete [] zvals;
+}
+
+void LocalEnvironment :: compute_indices(){
+
+    int no_atoms = rs.size();
+
+    // Initialize a list of lists storing atom indices.
+    int n_cutoffs = n_body_cutoffs.size();
```

Comment:
can't you just initialize n_body_indices and many_body_indices with an initial size?

Reply:
We would first need to count the atoms inside each cutoff sphere to get the size right. Which could certainly be done, and might be slightly faster, but I've instead opted to just push_back the list of indices while we loop over the atoms.